src/Hedgehog/Extras/Test/Base.hs

```diff
@@ -313,65 +324,76 @@ jsonErrorFailM f = f >>= jsonErrorFail
 -- | Run the operation 'f' once a second until it returns 'True' or the deadline expires.
 --
 -- Expiration of the deadline results in an assertion failure
-assertByDeadlineIO :: (MonadTest m, MonadIO m, HasCallStack) => UTCTime -> IO Bool -> m ()
-assertByDeadlineIO deadline f = GHC.withFrozenCallStack $ do
-  success <- liftIO f
-  unless success $ do
-    currentTime <- liftIO DTC.getCurrentTime
-    if currentTime < deadline
-      then do
-        liftIO $ IO.threadDelay 1000000
-        assertByDeadlineIO deadline f
-      else do
-        H.annotateShow currentTime
-        failMessage GHC.callStack "Condition not met by deadline"
+byDeadlineIO :: (MonadAssertion m, MonadTest m, MonadIO m, HasCallStack) => NominalDiffTime -> UTCTime -> IO a -> m a
+byDeadlineIO period deadline f = GHC.withFrozenCallStack $ byDeadlineM period deadline $ liftIO f
 
 -- | Run the operation 'f' once a second until it returns 'True' or the deadline expires.
 --
 -- Expiration of the deadline results in an assertion failure
-assertByDeadlineM :: (MonadTest m, MonadIO m, HasCallStack) => UTCTime -> m Bool -> m ()
-assertByDeadlineM deadline f = GHC.withFrozenCallStack $ do
-  success <- f
-  unless success $ do
-    currentTime <- liftIO DTC.getCurrentTime
-    if currentTime < deadline
-      then do
-        liftIO $ IO.threadDelay 1000000
-        assertByDeadlineM deadline f
-      else do
-        H.annotateShow currentTime
-        failMessage GHC.callStack "Condition not met by deadline"
+byDeadlineM :: forall m a. (MonadAssertion m, MonadTest m, MonadIO m, HasCallStack) => NominalDiffTime -> UTCTime -> m a -> m a
+byDeadlineM period deadline f = GHC.withFrozenCallStack $ do
+  start <- liftIO DTC.getCurrentTime
+  a <- goM
+  end <- liftIO DTC.getCurrentTime
+  note_ $ "Operation completed in " <> show (DTC.diffUTCTime end start)
+  return a
+  where goM :: m a
+        goM = H.catchAssertion f $ \e -> do
+          currentTime <- liftIO DTC.getCurrentTime
+          if currentTime < deadline
+            then do
+              liftIO $ IO.threadDelay (floor (DTC.nominalDiffTimeToSeconds period * 1000000))
```

Comment:
If `Pico` is 10^-12s shouldn't we multiple by 1 000 000 000 000?

Reply:
`threadDelay` takes microseconds, so the factor required is less.